# Analysis of Free Applications
In this project I will analyse free applications available in Apple Store.
Based on this anlysis I would like to recomend to my development team which applications they shoud create to get as much as possible sponsors with Ads

In [1]:
def explore_data(dataset, start, end, rows_and_columns = False):
    dataset_slice = dataset[start:end]
    rows = 0
    columns = 0
    for row in dataset_slice:
        print(row)
        print('\n')
        
    if rows_and_columns:
        print('Number of rows:', len(dataset))
        print('Number of columns:', len(dataset[0]))

In [2]:
import csv
openfile = open('AppleStore.csv', encoding='utf8')
handler = csv.reader(openfile)
datasetApple = list(handler)

openfile = open('googleplaystore.csv')
handler = csv.reader(openfile)
datasetGoogle = list(handler)

In [3]:
explore_data(datasetGoogle, 0, 2, True)
explore_data(datasetApple, 0, 2, True)

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver']


['Photo Editor & Candy Camera & Grid & ScrapBook', 'ART_AND_DESIGN', '4.1', '159', '19M', '10,000+', 'Free', '0', 'Everyone', 'Art & Design', 'January 7, 2018', '1.0.0', '4.0.3 and up']


Number of rows: 10842
Number of columns: 13
['id', 'track_name', 'size_bytes', 'currency', 'price', 'rating_count_tot', 'rating_count_ver', 'user_rating', 'user_rating_ver', 'ver', 'cont_rating', 'prime_genre', 'sup_devices.num', 'ipadSc_urls.num', 'lang.num', 'vpp_lic']


['284882215', 'Facebook', '389879808', 'USD', '0.0', '2974676', '212', '3.5', '3.5', '95.0', '4+', 'Social Networking', '37', '1', '29', '1']


Number of rows: 7198
Number of columns: 16


We have two data sets for Apple's and Google apps.
I downloaded it from bellow links:

[Apple Store](https://dq-content.s3.amazonaws.com/350/AppleStore.csv)

[Google Play](https://dq-content.s3.amazonaws.com/350/googleplaystore.csv)

Apple file contains 16 columns. For my analysis I use: 'track_name', 'price', 'prime_genre' and 'user_rating'

Google app file contains 13 columns. I use 'App', 'Genres', 'Rating', 'Reviews' and 'Price'

In [4]:
print(datasetGoogle[0], '\n', datasetGoogle[10473])

['App', 'Category', 'Rating', 'Reviews', 'Size', 'Installs', 'Type', 'Price', 'Content Rating', 'Genres', 'Last Updated', 'Current Ver', 'Android Ver'] 
 ['Life Made WI-Fi Touchscreen Photo Frame', '1.9', '19', '3.0M', '1,000+', 'Free', '0', 'Everyone', '', 'February 11, 2018', '1.0.19', '4.0 and up']


In [5]:
uniqueApps = []
nonUniqueApps = []

for row in datasetGoogle[1:]:
    name = row[0]
    if name in uniqueApps:
        nonUniqueApps.append(name)
    else:
        uniqueApps.append(name)
        
print(len(uniqueApps))
print(len(nonUniqueApps))


9660
1181


Based on discussion forum about this dataset I identified one row wiht wrong data. I decide to delte this row.

In [6]:
del datasetGoogle[10473]

In my detaset I found duplicate entries. I decided to remove duplicates based on number of reivews. This value is unique and is better for my project to anlayse data based on more reviews. 

I create empty directory, which will be hold name of application and number of reviews. I loop through dataset and look for this values

In [7]:
reviews_max = {}
for row in datasetGoogle[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if name in reviews_max and reviews_max[name] < n_reviews:
        reviews_max[name] = n_reviews
    elif name is not reviews_max:
        reviews_max[name] = n_reviews
        
print(len(reviews_max))

9659


Now I create two empty lists. android_clean will hold clean data set. "already_added" is control list to sotre names of apps which I add recently to "android_clean".

I loop through orginal data set, extract app name and number of reivews. If this numbers are match with dictionary, I record it in android_clean, additionaly application name should be not exist in our control list. In the same step I add app name to control list arleady_added.

In [8]:
android_clean = []
already_added = []

for row in datasetGoogle[1:]:
    name = row[0]
    n_reviews = float(row[3])
    if n_reviews == reviews_max[name] and name not in already_added:
        android_clean.append(row)
        already_added.append(name)

print(len(android_clean))

9659


Now I create function which is looking for non English characters in application names

In [9]:
def check_English(app_name):
    special_char = 0
    for character in app_name:
        if ord(character) > 127 and special_char > 2:
            return False
        elif ord(character) > 127:
            special_char += 1
    return True

In [10]:
check_English('Docs To Go™™™™ Free Office Suite')

False

In [11]:
index_id = []
for row in android_clean:
    if check_English(row[0]) != True:
        index_id.append(android_clean.index(row))
        
index_id.reverse()
for row in index_id:
    android_clean.pop(row)

In [12]:
for row in android_clean:
    if check_English(row[0]) != True:
        print("Do usunięcia")

In [13]:
index_id = []
for row in datasetApple:
    if check_English(row[1]) != True:
        index_id.append(datasetApple.index(row))
        
index_id.reverse()
for row in index_id:
    datasetApple.pop(row)
    
for row in datasetApple:
    if check_English(row[1]) != True:
        print("Do usnięcia")

I clean datasets from non Engish applications.
I list number of applications in each datasets.

In [14]:
print("Apple App Dataset length: ", len(datasetApple[1:]))
print("Android Dataset length: ", len(android_clean))

Apple App Dataset length:  6183
Android Dataset length:  9614


In [15]:
android_free = []
for row in android_clean:
    if row[6] == 'Free':
        android_free.append(row)

In [16]:
apple_free = []
for row in datasetApple[1:]:
    if float(row[4]) == 0.0:
        apple_free.append(row)

In [17]:
print("Apple Free Apps length: ", len(apple_free))
print("Android Free Apps length: ", len(android_free))

Apple Free Apps length:  3222
Android Free Apps length:  8863


I would like to find app profile which has good response from users on both markets.

I will analyse which genres are most common on each of markets. I will build frequency table for a few columns in our data sets.

In [18]:
datasetApple[0:2]

[['id',
  'track_name',
  'size_bytes',
  'currency',
  'price',
  'rating_count_tot',
  'rating_count_ver',
  'user_rating',
  'user_rating_ver',
  'ver',
  'cont_rating',
  'prime_genre',
  'sup_devices.num',
  'ipadSc_urls.num',
  'lang.num',
  'vpp_lic'],
 ['284882215',
  'Facebook',
  '389879808',
  'USD',
  '0.0',
  '2974676',
  '212',
  '3.5',
  '3.5',
  '95.0',
  '4+',
  'Social Networking',
  '37',
  '1',
  '29',
  '1']]

To genereate frequecny table and find most common genre in each market I identified columns which I should use to do this:

In Apple Store

rating_count_tot: 5

user_rating: 7

prime_genre: 11

In Google Play

Rating: 2

Reviews: 3

Genres: 9

Category: 1


In [19]:
def freq_table(dataset, index):
    genre_dict = {}
    for row in dataset:
        if row[index] in genre_dict:
            genre_dict[row[index]] += 1
        else:
            genre_dict[row[index]] = 1
    
    genre_freq = {}
    for genre_name in genre_dict:
        genre_freq[genre_name] = round(genre_dict[genre_name]/len(dataset) * 100, 2)
        
        
    return genre_freq

#test1 = freq_table(apple_free, 11)

In [20]:
def display_table(dataset, index):
    table = freq_table(dataset, index)
    table_display = []
    for key in table:
        key_val_as_tuple = (table[key], key)
        table_display.append(key_val_as_tuple)

    table_sorted = sorted(table_display, reverse = True)
    for entry in table_sorted:
        print(entry[1], ':', entry[0])

In [21]:
print("Frequency table for Apple Apps by Genre")
display_table(apple_free, 11)
print("")
print("Frequency table for Google Apps by Genre")
display_table(android_free, 9)
print("")
display_table(android_free, 1)


Frequency table for Apple Apps by Genre
Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12

Frequency table for Google Apps by Genre
Tools : 8.45
Entertainment : 6.07
Education : 5.35
Business : 4.58
Productivity : 3.89
Lifestyle : 3.89
Finance : 3.7
Medical : 3.54
Sports : 3.46
Personalization : 3.32
Communication : 3.25
Action : 3.1
Health & Fitness : 3.07
Photography : 2.94
News & Magazines : 2.8
Social : 2.66
Travel & Local : 2.32
Shopping : 2.25
Books & Reference : 2.14
Simulation : 2.04
Dating : 1.86
Arcade : 1.86
Video Players & Editors : 1.78
Casual : 1.75
Maps & Navigation : 1.4
Food & Drink : 1.24
Puzzle : 1.13
Racing : 0.99
Role Playing : 0.94

In [22]:
display_table(apple_free, 11)

Games : 58.16
Entertainment : 7.88
Photo & Video : 4.97
Education : 3.66
Social Networking : 3.29
Shopping : 2.61
Utilities : 2.51
Sports : 2.14
Music : 2.05
Health & Fitness : 2.02
Productivity : 1.74
Lifestyle : 1.58
News : 1.33
Travel : 1.24
Finance : 1.12
Weather : 0.87
Food & Drink : 0.81
Reference : 0.56
Business : 0.53
Book : 0.43
Navigation : 0.19
Medical : 0.19
Catalogs : 0.12


Most common genre in App Store is Games which is 58.16% of all apps. Entertainment and Photo & Video are most popular too.
Most apps are disigned for games, photo and video. 

We can recomend for App Store Games genre application.

In [23]:
display_table(android_free, 1)

FAMILY : 19.21
GAME : 9.51
TOOLS : 8.46
BUSINESS : 4.58
LIFESTYLE : 3.9
PRODUCTIVITY : 3.89
FINANCE : 3.7
MEDICAL : 3.54
SPORTS : 3.42
PERSONALIZATION : 3.32
COMMUNICATION : 3.25
HEALTH_AND_FITNESS : 3.07
PHOTOGRAPHY : 2.94
NEWS_AND_MAGAZINES : 2.8
SOCIAL : 2.66
TRAVEL_AND_LOCAL : 2.34
SHOPPING : 2.25
BOOKS_AND_REFERENCE : 2.14
DATING : 1.86
VIDEO_PLAYERS : 1.78
MAPS_AND_NAVIGATION : 1.4
FOOD_AND_DRINK : 1.24
EDUCATION : 1.13
LIBRARIES_AND_DEMO : 0.94
AUTO_AND_VEHICLES : 0.93
ENTERTAINMENT : 0.88
HOUSE_AND_HOME : 0.82
WEATHER : 0.8
EVENTS : 0.71
PARENTING : 0.65
ART_AND_DESIGN : 0.64
COMICS : 0.62
BEAUTY : 0.6


For Android apps most popular genre is Family. Games and Tools are the next popular type of applications. We observ similar situation like in Apple application. Genre is from games, entertaiment. 

I am recomending Family genre app.

In general after this analysis we can say that fun application are most frequently used.

In [24]:
freq_table(apple_free, 11)

{'Social Networking': 3.29,
 'Photo & Video': 4.97,
 'Games': 58.16,
 'Music': 2.05,
 'Reference': 0.56,
 'Health & Fitness': 2.02,
 'Weather': 0.87,
 'Utilities': 2.51,
 'Travel': 1.24,
 'Shopping': 2.61,
 'News': 1.33,
 'Navigation': 0.19,
 'Lifestyle': 1.58,
 'Entertainment': 7.88,
 'Food & Drink': 0.81,
 'Sports': 2.14,
 'Book': 0.43,
 'Finance': 1.12,
 'Education': 3.66,
 'Productivity': 1.74,
 'Business': 0.53,
 'Catalogs': 0.12,
 'Medical': 0.19}

In [25]:
freq_table_apple = freq_table(apple_free, 11)

In [26]:
freq_table_apple

{'Social Networking': 3.29,
 'Photo & Video': 4.97,
 'Games': 58.16,
 'Music': 2.05,
 'Reference': 0.56,
 'Health & Fitness': 2.02,
 'Weather': 0.87,
 'Utilities': 2.51,
 'Travel': 1.24,
 'Shopping': 2.61,
 'News': 1.33,
 'Navigation': 0.19,
 'Lifestyle': 1.58,
 'Entertainment': 7.88,
 'Food & Drink': 0.81,
 'Sports': 2.14,
 'Book': 0.43,
 'Finance': 1.12,
 'Education': 3.66,
 'Productivity': 1.74,
 'Business': 0.53,
 'Catalogs': 0.12,
 'Medical': 0.19}

In [27]:
for genre in freq_table_apple:
    total = 0
    len_genre = 0
    for row in apple_free:
        genre_app = row[11]
        if genre_app == genre:
            total += float(row[5])
            len_genre += 1
            
    avg_user_rating = total/len_genre
    print(genre, ": ", avg_user_rating)


Social Networking :  71548.34905660378
Photo & Video :  28441.54375
Games :  22788.6696905016
Music :  57326.530303030304
Reference :  74942.11111111111
Health & Fitness :  23298.015384615384
Weather :  52279.892857142855
Utilities :  18684.456790123455
Travel :  28243.8
Shopping :  26919.690476190477
News :  21248.023255813954
Navigation :  86090.33333333333
Lifestyle :  16485.764705882353
Entertainment :  14029.830708661417
Food & Drink :  33333.92307692308
Sports :  23008.898550724636
Book :  39758.5
Finance :  31467.944444444445
Education :  7003.983050847458
Productivity :  21028.410714285714
Business :  7491.117647058823
Catalogs :  4004.0
Medical :  612.0


Most popular genre base on number of reatings is Navigation app which is only 0.19% of frequecy usage.

In [28]:
for app in apple_free:
    if app[11] == 'Navigation':
        print(app[1], app[5])

Waze - GPS Navigation, Maps & Real-time Traffic 345046
Google Maps - Navigation & Transit 154911
Geocaching® 12811
CoPilot GPS – Car Navigation & Offline Maps 3582
ImmobilienScout24: Real Estate Search in Germany 187
Railway Route Search 5


In [29]:
display_table(android_free, 5)

1,000,000+ : 15.75
100,000+ : 11.56
10,000,000+ : 10.5
10,000+ : 10.21
1,000+ : 8.39
100+ : 6.92
5,000,000+ : 6.83
500,000+ : 5.56
50,000+ : 4.77
5,000+ : 4.51
10+ : 3.54
500+ : 3.25
50,000,000+ : 2.3
100,000,000+ : 2.13
50+ : 1.92
5+ : 0.79
1+ : 0.51
500,000,000+ : 0.27
1,000,000,000+ : 0.23
0+ : 0.05


In [30]:
freq_table_android = freq_table(android_free, 1)

In [31]:
freq_table_android

{'ART_AND_DESIGN': 0.64,
 'AUTO_AND_VEHICLES': 0.93,
 'BEAUTY': 0.6,
 'BOOKS_AND_REFERENCE': 2.14,
 'BUSINESS': 4.58,
 'COMICS': 0.62,
 'COMMUNICATION': 3.25,
 'DATING': 1.86,
 'EDUCATION': 1.13,
 'ENTERTAINMENT': 0.88,
 'EVENTS': 0.71,
 'FINANCE': 3.7,
 'FOOD_AND_DRINK': 1.24,
 'HEALTH_AND_FITNESS': 3.07,
 'HOUSE_AND_HOME': 0.82,
 'LIBRARIES_AND_DEMO': 0.94,
 'LIFESTYLE': 3.9,
 'GAME': 9.51,
 'FAMILY': 19.21,
 'MEDICAL': 3.54,
 'SOCIAL': 2.66,
 'SHOPPING': 2.25,
 'PHOTOGRAPHY': 2.94,
 'SPORTS': 3.42,
 'TRAVEL_AND_LOCAL': 2.34,
 'TOOLS': 8.46,
 'PERSONALIZATION': 3.32,
 'PRODUCTIVITY': 3.89,
 'PARENTING': 0.65,
 'WEATHER': 0.8,
 'VIDEO_PLAYERS': 1.78,
 'NEWS_AND_MAGAZINES': 2.8,
 'MAPS_AND_NAVIGATION': 1.4}

In [32]:
for category in freq_table_android:
    total = 0
    len_category = 0
    for row in android_free:
        category_app = row[1]
        if category_app == category:
            number_of_installs = row[5]
            n_installs = number_of_installs.replace('+', '')
            n_installs = n_installs.replace(',', '')
            total += float(n_installs)
            len_category += 1
    avg_n_installs = total / len_category
    print(category, ": ", round(avg_n_installs, 1))

ART_AND_DESIGN :  1986335.1
AUTO_AND_VEHICLES :  647317.8
BEAUTY :  513151.9
BOOKS_AND_REFERENCE :  8767811.9
BUSINESS :  1704192.3
COMICS :  817657.3
COMMUNICATION :  38326063.2
DATING :  854028.8
EDUCATION :  1768500.0
ENTERTAINMENT :  9146923.1
EVENTS :  253542.2
FINANCE :  1387692.5
FOOD_AND_DRINK :  1924897.7
HEALTH_AND_FITNESS :  4167457.4
HOUSE_AND_HOME :  1331540.6
LIBRARIES_AND_DEMO :  638503.7
LIFESTYLE :  1437816.3
GAME :  12914435.9
FAMILY :  5183203.6
MEDICAL :  123064.8
SOCIAL :  23253652.1
SHOPPING :  7036877.3
PHOTOGRAPHY :  17840110.4
SPORTS :  4274688.7
TRAVEL_AND_LOCAL :  13984077.7
TOOLS :  10801391.3
PERSONALIZATION :  5201482.6
PRODUCTIVITY :  16772838.6
PARENTING :  542603.6
WEATHER :  5074486.2
VIDEO_PLAYERS :  24790074.2
NEWS_AND_MAGAZINES :  9549178.5
MAPS_AND_NAVIGATION :  4056941.8


Most recomended app genre based on number of install is Communication